File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.js
/**
 * @param {string} s
 * @return {number}
 */
var longestPalindromeSubseq = function(s) {
  const n = s.length;
  const dp = new Array(n).fill(0);
  
  for (let i = n - 1; i >= 0; i--) {
    dp[i] = 1;
    let prev = 0;
    for (let j = i + 1; j < n; j++) {
      let temp = dp[j];
      dp[j] = Math.max(dp[j], dp[j - 1], prev + (s[i] === s[j] ? 2 : 0));
      prev = temp;
    }
  }
  
  return dp[n - 1];
};

File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        dp = [0] * n
        for i in range(n - 1, -1, -1):
            dp[i], prev = 1, 0
            for j in range(i + 1, n):
                tmp = dp[j]
                if s[i] == s[j]:
                    prev += 2
                dp[j] = max(dp[j], dp[j - 1], prev)
                prev = tmp
        return dp[-1]